await-generator/src/SOFe/AwaitGenerator/Loading.php

```diff
@@ -63,9 +63,14 @@ public function __construct(Closure $loader){
 	 */
 	public function get() : Generator{
 		if($this->onLoaded !== null){
-			yield from Await::promise(function($resolve){
-				$this->onLoaded[] = $resolve;
-			});
+			$key = rand();
```

Comment:
Why not `len`? I believe `rand()` slows down the process a lot. (Besides the latency of PHP's alias link from `rand()` to `mt_rand()`, also the latency of calling a thread rand gen in general.)

Reply:
Please `use function rand;`

Reply:
True, I could use a static int counter indeed. Wonder why I didn't think it that.

Reply:
> However this counter needs to be wrapping so that it doesn't get stuck at `(float) PHP_INT_MAX`.

But PHP array does not support float at all: https://3v4l.org/4it7M

Or do you mean using a `list<list<...>>`?

Reply:
@Endermanbugzjfc the counter will be stuck at a float, that's why I want to make it a wrapping integer instead.

Reply:
> @Endermanbugzjfc the counter will be stuck at a float, that's why I want to make it a wrapping integer instead.

Actually, why don't we `(string)` the float? It should work as long as `strict_types=1`.

Reply:
@Endermanbugzjfc because float precision. 
```console
$ php -r '$a = (float) PHP_INT_MAX; $b = $a + 1; var_dump($a === $b);'
bool(true)
```

Reply:
Anyway, I changed this from `rand()` to `bcadd()` counter. Not sure about the performance, but at least we are sure it will never duplicate, although this counter might grow indefinitely (which is still a leak in the O(log t) sense, but who cares anyway).

Reply:
Wait, why do I even need to make this counter static... It could have worked well with just a local counter, and since `PHP_INT_MAX >= count($anyArray)` must always be true, we don't need to worry about integer wrapping for a single object.

We still have the chance of integer wrapping if a `Loading` never resolves but someone called `yield from $loading->get()` *and* cancelled it for more than `PHP_INT_MAX` times, but I suppose we don't care anyway...

Reply:
Never mind, the whole thing is flawed to begin with. I could just have used `spl_object_id($resolve)` as the key.